libs/db/boltdb.go

```diff
@@ -0,0 +1,313 @@
+		})
+		return putErr
+	})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func (bdb *boltDBBatch) WriteSync() {
+	bdb.Write()
+}
+
+func (bdb *boltDBBatch) Close() {}
+
+// WARNING: Any concurrent writes (Set, SetSync) will block until the Iterator
+// is closed.
+func (bdb *BoltDB) Iterator(start, end []byte) Iterator {
+	tx, err := bdb.db.Begin(false)
+	if err != nil {
+		panic(err)
+	}
+	c := tx.Bucket(bucket).Cursor()
+	return newBoltDBIterator(c, start, end, false)
+}
+
+// WARNING: Any concurrent writes (Set, SetSync) will block until the Iterator
+// is closed.
+func (bdb *BoltDB) ReverseIterator(start, end []byte) Iterator {
+	tx, err := bdb.db.Begin(false)
+	if err != nil {
+		panic(err)
+	}
+	c := tx.Bucket(bucket).Cursor()
+	return newBoltDBIterator(c, start, end, true)
+}
+
+// boltDBIterator allows you to iterate on range of keys/values given some
+// start / end keys (nil & nil will result in doing full scan).
+type boltDBIterator struct {
+	itr   *bbolt.Cursor
+	start []byte
+	end   []byte
+
+	currentKey   []byte
+	currentValue []byte
+
+	isInvalid bool
+	isReverse bool
+}
+
+func newBoltDBIterator(itr *bbolt.Cursor, start, end []byte, isReverse bool) *boltDBIterator {
+	var ck, cv []byte
+	if isReverse {
+		if end == nil {
+			ck, cv = itr.Last()
+		} else {
+			_, _ = itr.Seek(end) // after key
+			ck, cv = itr.Prev()  // return to end key
+		}
+	} else {
+		if start == nil {
+			ck, cv = itr.First()
+		} else {
+			ck, cv = itr.Seek(start)
+		}
+	}
+
+	return &boltDBIterator{
+		itr:          itr,
+		start:        start,
+		end:          end,
+		currentKey:   ck,
+		currentValue: cv,
+		isReverse:    isReverse,
+		isInvalid:    false,
+	}
+}
+
+func (itr *boltDBIterator) Domain() ([]byte, []byte) {
+	return itr.start, itr.end
+}
+
+func (itr *boltDBIterator) Valid() bool {
+	if itr.isInvalid {
+		return false
+	}
+
+	// iterated to the end of the cursor
+	if len(itr.currentKey) == 0 {
+		itr.isInvalid = true
+		return false
+	}
+
+	if itr.isReverse {
+		if itr.start != nil && bytes.Compare(itr.currentKey, itr.start) < 0 {
+			itr.isInvalid = true
+			return false
+		}
+	} else {
+		if itr.end != nil && bytes.Compare(itr.end, itr.currentKey) <= 0 {
+			itr.isInvalid = true
+			return false
+		}
+	}
+
+	// Valid
+	return true
+}
+
+func (itr *boltDBIterator) Next() {
+	itr.assertIsValid()
+	if itr.isReverse {
+		itr.currentKey, itr.currentValue = itr.itr.Prev()
+	} else {
+		itr.currentKey, itr.currentValue = itr.itr.Next()
+	}
+}
+
+func (itr *boltDBIterator) Key() []byte {
+	itr.assertIsValid()
+	return itr.currentKey
+}
+
+func (itr *boltDBIterator) Value() []byte {
+	itr.assertIsValid()
+	return itr.currentValue
+}
+
+// boltdb cursor has no close op.
+func (itr *boltDBIterator) Close() {}
+
+func (itr *boltDBIterator) assertIsValid() {
+	if !itr.Valid() {
+		panic("Boltdb-iterator is invalid")
+	}
+}
+
+// nonEmptyKey returns a []byte("nil") if key is empty.
+// WARNING: this may collude with "nil" user key!
```

Comment:
bolt does not support empty keys!

Reply:
I think that we should judge if the key is empty.  
If the key is empty,  we should return an error (key must not be empty)
In fact, many embedded KVDBs do not allow key to be empty.

We should modify the nonEmptyKey() that return an error when key is empty.

Reply:
Are we using non-empty keys anywhere (the sdk)? If not, returning an error makes sense. Otherwise, we would need to support existing use-cases nonetheless.

Reply:
> Are we using non-empty keys anywhere (the sdk)?

This logic was explicitly added by Jae. Not sure what for. Maybe for SDK?

Reply:
OK, then let's keep it like you've implemented it (for now).

Reply:
I want to know where to use empty-key in cosmos-sdk.  
`[]byte("nil")` is not a good idea because you probably judge the key in other places.   
And if you let the empty-key keep it ,  it will be incompatible with other better KVDBs.

Reply:
cc @alexanderbez @jaekwon

Reply:
> Ethan: also how do we handle the empty key (non nil) on the fsdb?
> Jae: mmm lets panic for fsdb.

Reply:
4ce8448d7fcf92b040046f894474ce2f7e779b67

Reply:
https://github.com/tendermint/tendermint/issues/3614